### ISSUE TITLE
Neos:///session links do not let you connect to hidden Sessions. 

### ISSUE THREAD

**Issue by TehTurk** [Neos-Metaverse/NeosPublic#2494]
## Describe the bug
Currently to link up some sort of instancing between worlds, it's tricky. Your required to get the lnl link or steam link to actually connect.(But to get one you need to connect first) (WebLinks do not register as well). This leads me to be a bug as if lnl and steam work, neos:///session and web links should work too. 

## To Reproduce
1. make Headless hidden instance
2. have preset id to connect to
3. Unless you connect directly, Neos will no find the session and tou get the lnl/steam link directly. 

## Expected behavior
A preset ID should work for situations as this. Or if folks want a Session to be an open invite link only. 

## Screenshots / Video
[If applicable, add screenshots or a video to help explain your problem. DO NOT use screenshots or video in place of description of the problem and replication steps though.]

## Bug information (please complete the following information):
 - Always 
 - Yes
- Current

## Additional context
This was a partial request on VBLFC as this was to make their onboarding experiences smoother.

## Reporters:
Turk, Gawdly

**Comment by Frooxius** [comment]
This is a known issue and a bit tricky one to handle. The Session ID needs to be resolved to actual networking URL's, which is done via the cloud.

However hidden sessions do not send anything to the cloud out of principle, meaning it's not possible to resolve those links.

There are two main approaches for this:
- Have hidden sessions only send the URL's and no other data. However this opens up a security hole, if someone gets this data, they'll be able to connect to the session unless you have some other mechanism in place.
- Encode all the URL's as part of the session URL. This can be trickier and can potentially result in very big URLs. 

I'm not sure which one would be better to take here.

**Comment by TehTurk** [comment]
I'd say the former. If a link is gotten from somewhere it's typically on the onus of the user as someone got the session and sent it out. Or, someone wasn't clear on their procedure of sharing the link to the world. 

Hmm it could work but is there no way to have a custom format be encoded?

**Comment by shiftyscales** [comment]
This is essentially a duplicate of #922. Given the age, and lack of activity/detail on that particular issue, I'll close it in favor of this one.

**Comment by TehTurk** [comment]
Following up from #2631 some other reasons were if a user doesn't have any sort of contact with an individual it's tricky to get them into the world(Only through invite). I can see the former solution being a benefit for private events, or community only events where the link is required.

**Comment by Hayden-da-Serg** [comment]
This became especially prevalent one night when the message history between contacts did not want to load, the only way of getting into a hidden session was by giving out the `lnl-nat` or `neos-steam` URL and instructing the user to insert the URL into a session orb, which is not at all ideal.

**Comment by Frooxius** [comment]
If there are cloud issues with things not loading, then this solution won't unfortunately help you, as that requires the cloud to resolve the actual URL's.

We might encode the URL's into the main one, but it'll also make it pretty long. I'm not sure if that'd be acceptable solution or not.

However you should be able to click on the lnl-nat and neos-steam URL's, Neos registers itself to handle those.

**Comment by Hayden-da-Serg** [comment]
I was considering making a GitHub issue about allowing lnl-nat or neos-steam URLs to be pasted into the Neos window to attempt a connection. In my experience the only way I could personally use them is by putting them into a session orb.

**Comment by TehTurk** [comment]
> If there are cloud issues with things not loading, then this solution won't unfortunately help you, as that requires the cloud to resolve the actual URL's.
> 
> We might encode the URL's into the main one, but it'll also make it pretty long. I'm not sure if that'd be acceptable solution or not.
> 
> However you should be able to click on the lnl-nat and neos-steam URL's, Neos registers itself to handle those.

I think I'd avoid against the big URLs. While I understand Hidden is hidden from cloud and doesn't send data, I think allowing things like the session links to resolve would be enough imho. StaticURLs kinda help with this, but oft not if a hidden session doesn't have a static url, it's session url is used as a static link on the fly. Only way to get into that session would be if someone from inside that session shares the link (Which makes sense in a security sense) If your already in, at that point extra security has no point.

**Comment by Frooxius** [comment]
> 
> 
> > If there are cloud issues with things not loading, then this solution won't unfortunately help you, as that requires the cloud to resolve the actual URL's.
> > We might encode the URL's into the main one, but it'll also make it pretty long. I'm not sure if that'd be acceptable solution or not.
> > However you should be able to click on the lnl-nat and neos-steam URL's, Neos registers itself to handle those.
> 
> I think I'd avoid against the big URLs. While I understand Hidden is hidden from cloud and doesn't send data, I think allowing things like the session links to resolve would be enough imho. StaticURLs kinda help with this, but oft not if a hidden session doesn't have a static url, it's session url is used as a static link on the fly. Only way to get into that session would be if someone from inside that session shares the link (Which makes sense in a security sense) If your already in, at that point extra security has no point.

I'm sorry, but I don't quite understand what do you mean? What are StaticURLs in this context? Session URL doesn't change over its lifetime. 

This solution doesn't really have much to do with security. If you have the URL you'll be able to get the actual URLs to connect to either way, it's just which method will be better.

Encoding the protocol URL's into the session one would work without any dependence on the cloud to resolve it at all and could be more robust, only thing is that the URL itself would end up substantially longer, which might be a bit unsightly when pasting it.

@Hayden-da-Serg You might need to click the URL in external program or web browser. Not everything recognizes the protocols though, so you might need to use this one instead: https://api.neos.com/open/session/lnl-nat/<id>

**Comment by Hayden-da-Serg** [comment]
Even when entering the lnl URLs directly into the address bar or even the Run program they give nothing, through my browser it just goes to search for it, through Run it brings up the Windows Open With dialogue.

**Comment by Aerizeon** [comment]
For me, I usually get an error from Firefox for Session URLs and such (when using the mentor signal, for example):
![Image](https://media.discordapp.net/attachments/750741439880888432/863617324258557962/unknown.png)

**Comment by TehTurk** [comment]
> I'm sorry, but I don't quite understand what do you mean? What are StaticURLs in this context? Session URL doesn't change over its lifetime.
> 
> This solution doesn't really have much to do with security. If you have the URL you'll be able to get the actual URLs to connect to either way, it's just which method will be better.
> 
> Encoding the protocol URL's into the session one would work without any dependence on the cloud to resolve it at all and could be more robust, only thing is that the URL itself would end up substantially longer, which might be a bit unsightly when pasting it.
> 
> @Hayden-da-Serg You might need to click the URL in external program or web browser. Not everything recognizes the protocols though, so you might need to use this one instead: https://api.neos.com/open/session/lnl-nat/your_id_here

Ah the Static Session urls like S-U-MechaTurk:Metaverse_Hub. With the changing I was speaking more so with how from session to session, (Say if session crashes, or a team is working on a project) is more so what I meant where the copied url is used as a temporary link and sometimes can be used for quite a bit.

Hmm I see. Do you have examples? I Know discord has invite links as well but those can be pretty similar or lengthy in nature.

**Comment by Gawdl3y** [comment]
I agree that URL length isn't really an issue, especially if you can just use the paste content from clipboard button to join a session/spawn a world orb for it. That being said, what *would* be an issue, is the URL not being able to be static. If the LNL/Steam URLs were encoded in the session URL, wouldn't that result in it changing with each restart, thus defeating the purpose of the static session IDs?

Personally, I see no issue with the cloud having knowledge of hidden sessions with static session IDs. As long as it's not being presented to users/API responses in any kind of world listing, it's achieved the goal of the "hidden" status in my mind. I understand the hesitation around that, however, as any bug resulting in those URLs being sent along to someone could potentially be harmful to those running the hidden sessions at some point.

I think the biggest problem is that there's simply no convenient way to run a persistent session that is accessible only to people with knowledge of it currently. You can run a headless set to contacts only, but then people can't freely invite others, and you have to manually wrangle contact requests on the headless user. You can run a hidden session, but then you can't have it be persistent/always accessible from the same URL.

Ultimately, I think that the most ideal capability would be to have a session behind a whitelist of some kind (using those cloud variables, probably), optionally allowing invites, and automatically show up in the world browser when whitelisted/invited, allowing joining like a normal session. Thanks to discussions I've had with staff members, however, I know that this would require a pretty big restructure of the way the world browser works, and thus probably isn't realistic in the near-term.

**Comment by Frooxius** [comment]
I see. Yeah if you need the URL to remain the same without having to refresh, then we'll have to go with the cloud solution.

The LNL URL generally stays the same with custom session ID, but the Steam one does not, so that wouldn't work unfortunately.

I'll see if we can have the Hidden session send the URL's to the cloud. That should be good enough compromise.

For it appearing in the world browser, that's tricky. A lot of the validation happens locally on the host, so the cloud isn't necessarily aware of it. The showing of session and validating join are separate things and as we add more complex join rules, it'll be harder for the cloud to determine who can see the session and who can't.

It also opens up some potential issues that Hidden sessions are meant to solve too, where it could appear in places where you don't want it to.
